### PR TITLE
Add support for torch.gather OP (backend implementation)

### PR DESCRIPTION
**Pull request by ftxj**
* add torch_gather(lookup_tv, dim, index_tv) fusion for nvfuser: 1) support arbitrarily rank; 2) support arbitrarily dim value; 3) index_tv can be the result of another operator, but input_tv cannot. 4) only support input_tv in global memory

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/ir_internal_nodes.h [hunk 1]
```diff
@@ -103,6 +103,41 @@ class TORCH_CUDA_CU_API IndexSelectOp : public Expr {
...
+
+  int dim() const {
+    return attribute(0)->as<Attribute<int>>()->value;
+  }
+
+  Val* out() const {
+    return output(0);
+  }
+
+  Val* in1() const {
```
Comment (azazhu) [inline torch/csrc/jit/codegen/cuda/ir_internal_nodes.h]: out()/in1()/in2() can be removed

File: torch/csrc/jit/codegen/cuda/arith.h [hunk 2]
```diff
@@ -553,6 +553,13 @@ TORCH_CUDA_CU_API TensorView* index_select(
     TensorView* input,
     int dim,
     TensorView* index);
+// torch.gather
+TORCH_CUDA_CU_API TensorView* torch_gather(
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/arith.h:557]: We may want to rename the existing nvfuser gather to something else, but that can be done later.

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp [hunk 3]
```diff
@@ -551,6 +551,17 @@ void IrPrinter::handle(const IndexSelectOp* sop) {
   os_ << ", dim = " << sop->dim() << ", " << sop->input(1) << " )\n";
 }
 
+void IrPrinter::handle(const TorchGatherOp* gop) {
+  indent() << gop->output(0) << "\n";
+  indent() << "   = gather( ";
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_iostream.cpp]: `torch_gather` at least for now

File: torch/csrc/jit/codegen/cuda/type.h [hunk 4]
```diff
@@ -305,6 +305,7 @@ enum class IterType {
   Broadcast,
   Gather,
   Stride,
+  GatherScatter,
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/type.h:311]: Why is this necessary?

File: torch/csrc/jit/codegen/cuda/lower_replace_size.cpp [hunk 5]
```diff
@@ -40,6 +40,10 @@ std::unordered_map<Val*, Val*> getSimplificationMap(Fusion* fusion) {
     if (id0->isBroadcast() || id1->isBroadcast()) {
       return;
     }
+    
+    if (id0->isGatherScatter() || id1->isGatherScatter()) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_replace_size.cpp:44]: Is this the only place `IterType::GatherScatter` is used?

**Review by naoyam — COMMENTED** [review]
Just some quick comments for now

File: torch/csrc/jit/codegen/cuda/lower_replace_size.cpp (hunk 5 above)
Comment (ftxj) [reply torch/csrc/jit/codegen/cuda/lower_replace_size.cpp:44]: Until now, yes. And this new iteration type handles the case where the dimension sizes of consumer and producer are not equal.

File: torch/csrc/jit/codegen/cuda/lower_index.cpp [hunk 6]
```diff
@@ -221,6 +221,83 @@ void IndexLowering::handle(const IndexSelectOp* sop) {
   GpuLower::current()->propagateExprInfo(sop, back());
 }
 
+namespace {
+Val* packToRealIndices(kir::TensorIndex* idx) {
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/lower_index.cpp]: Move this to `index_compute.cpp`

File: torch/csrc/jit/codegen/cuda/lower_index.cpp [hunk 7]
```diff
@@ -221,6 +221,83 @@ void IndexLowering::handle(const IndexSelectOp* sop) {
...
+        index_domains[dim]->extent() 
+      ), 
+      input_tv_strides[dim + 1]
+    );
+    input_tv_index.push_back(index_tv_index_for_dim);
+  }
+
+  input_tv_index[top->dim()] = IrBuilder::mulExpr(indices, input_tv_strides[top->dim() + 1]);
+  auto input = SimplifyingIrBuilder::create<kir::TensorIndex>(
+      dynamic_cast<TensorView*>(top->input(0)), input_tv_index);
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/lower_index.cpp]: Refactor this and move it to a separate function, like `Index::getGatherGlobalStridedIndex`

File: torch/csrc/jit/codegen/cuda/ir_utils.cpp [hunk 8]
```diff
@@ -753,6 +765,30 @@ bool isIndexSelectIndicesTv(const TensorView* tv) {
   return false;
 }
 
+bool isTorchGatherLookupTv(const Val* tv) {
+  for (auto expr : tv->uses()) {
+    if (expr->isA<TorchGatherOp>()) {
+      auto idx_sel = expr->as<TorchGatherOp>();
+      if (idx_sel->input(0) == tv) {
```
Comment (yangjunpro) [inline torch/csrc/jit/codegen/cuda/ir_utils.cpp]: Just curious,  when `tv->uses()` tells that the expr is `TorchGatherOp`, it seems that there is already enough information to `return true` here? Why do we add another `idx_sel->input(0) == tv)` check here?

Comment (yangjunpro) [reply torch/csrc/jit/codegen/cuda/ir_utils.cpp]: Aha, got it, it is used to distinguish between different argument of `TorchGather`, ignore pls.

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 9]
```diff
@@ -355,6 +357,10 @@ TORCH_CUDA_CU_API bool isIndexSelectLookupTv(const TensorView* tv);
 // Check if the given tv is third argment of index_select(lookup, dim, indices)
 TORCH_CUDA_CU_API bool isIndexSelectIndicesTv(const TensorView* tv);
 
+// Check if the given tv is first/third argment of torch_gather(lookup, dim, indices)
```
Comment (yangjunpro) [inline torch/csrc/jit/codegen/cuda/ir_utils.h]: If `first/third`, then why in the above `isTorchGatherIndicesTv/isTorchGatherLookupTv`, we use `input(0)/input(1)` rather than `input(0)/input(2)`?

File: torch/csrc/jit/codegen/cuda/lower_index.cpp [hunk 10]
```diff
@@ -221,6 +221,83 @@ void IndexLowering::handle(const IndexSelectOp* sop) {
...
+Val* packToRealIndices(kir::TensorIndex* idx) {
+  Val* real_indices = nullptr;
+  bool first = true;
+  for(auto x : idx->indices()) {
+    if(first) {
+      real_indices = x;
+      first = false;
+    }
+    else {
+      real_indices = SimplifyingIrBuilder::addExpr(real_indices, x);
```
Comment (yangjunpro) [inline torch/csrc/jit/codegen/cuda/lower_index.cpp]: Nit. A compact way to do this may be 
`real_indices = SimplifyingIrBuilder::addExpr( real_indices, x)` since the underneath implementation of `addExpr` will handle the corresponding condition logic
image

File: torch/csrc/jit/codegen/cuda/lower_index.cpp [hunk 11]
```diff
@@ -221,6 +221,83 @@ void IndexLowering::handle(const IndexSelectOp* sop) {
...
+    }
+  }
+  return real_indices;
+}
+
+std::vector<Val*> getStrideForEachAxis(std::vector<IterDomain*> &domain) {
+  std::vector<Val*> strides(1, IrBuilder::create<Int>(1));
+  Val* stride = IrBuilder::create<Int>(1);
```
Comment (yangjunpro) [inline torch/csrc/jit/codegen/cuda/lower_index.cpp]: It seems the `stride` object created in 242 line is never pushed into the `strides` container?

File: torch/csrc/jit/codegen/cuda/lower_index.cpp [hunk 12]
```diff
@@ -221,6 +221,83 @@ void IndexLowering::handle(const IndexSelectOp* sop) {
...
+  for(auto x = domain.rbegin(); x != domain.rend(); ++x) {
+    stride = SimplifyingIrBuilder::mulExpr(stride, (*x)->extent());
+    strides.push_back(stride);
+  }
+  std::reverse(strides.begin(), strides.end());
```
Comment (yangjunpro) [inline torch/csrc/jit/codegen/cuda/lower_index.cpp]: Why do we traverse the `domain` in a reverse order, then do a `reverse` operation on `strides` object? Maybe we can get the same result by just traversing the `domain` in the ordinary order?

File: torch/csrc/jit/codegen/cuda/lower_index.cpp [hunk 13]
```diff
@@ -221,6 +221,83 @@ void IndexLowering::handle(const IndexSelectOp* sop) {
...
+  return strides;
+}
+}
+
+void IndexLowering::handle(const TorchGatherOp* top) {
+
+  const auto indices = lowerSrcIndex(top->input(1), top->output(0));
+  // const std::unordered_map<IterDomain*, Val*> override_index;
```
Comment (yangjunpro) [inline torch/csrc/jit/codegen/cuda/lower_index.cpp]: Remove it.

File: torch/csrc/jit/codegen/cuda/lower_index.cpp (hunk 7 above)
Comment (yangjunpro) [inline torch/csrc/jit/codegen/cuda/lower_index.cpp]: Why do we need to introduce a `dynamic_cast` here?

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp [hunk 14]
```diff
@@ -100,6 +100,7 @@ std::unordered_map<IterDomain*, IterDomain*> PairwiseRootDomainMap::map(
   }
 
   std::unordered_map<IterDomain*, IterDomain*> dom_map;
+
```
Comment (yangjunpro) [inline torch/csrc/jit/codegen/cuda/root_domain_map.cpp]: Why do we touch this line?

File: torch/csrc/jit/codegen/cuda/scheduler/pointwise_utils.cpp [hunk 15]
```diff
@@ -15,6 +15,13 @@ DomainMap::DomainMap(Fusion* fusion) : fusion_(fusion), ca_map_(fusion) {
   for (auto select : ir_utils::getIndexSelectOps(fusion)) {
     select_ids_.emplace(select->getSelectAxis());
   }
+  for (auto select : ir_utils::getTorchGatherOps(fusion)) {
+    auto inp = dynamic_cast<TensorView*>(select->input(0));
```
Comment (yangjunpro) [inline torch/csrc/jit/codegen/cuda/scheduler/pointwise_utils.cpp]: Ditto. Why do we need to introduce a `dynamic_cast` here?

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp [hunk 16]
```diff
@@ -34,6 +34,35 @@ std::deque<std::deque<TensorView*>> tvChains(
...
+  if (torch_gather->input(0)->uses().size() > 1) {
+    scheduler_debug_utils::canScheduleRejectReason(
+      schedule_stragety,
+      "First input of TorchGatherOp can only be used by TorchGatherOp");
+      return true;
+  }
+  if (!torch_gather->input(1)->isFusionInput()) {
+    scheduler_debug_utils::canScheduleRejectReason(
+      schedule_stragety,
+      "Index input of TorchGatherOp must be fusion input.");
```
Comment (yangjunpro) [inline torch/csrc/jit/codegen/cuda/scheduler/registry.cpp]: Nit. Error message is inconsistent. In the above it is mentioned `first input`, here we mention`Index input`.

File: torch/csrc/jit/codegen/cuda/test/test_gpu_gather_ops.cpp [hunk 17]
```diff
@@ -0,0 +1,278 @@
+#if defined(USE_CUDA)
+#include <gmock/gmock-matchers.h>
+#include <gtest/gtest.h>
+
```
Comment (yangjunpro) [inline torch/csrc/jit/codegen/cuda/test/test_gpu_gather_ops.cpp:4]: Such a large set of `#include`, are we for sure that all the included files are necessary?

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 9 above)
Comment (ftxj) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: The API interface is `torch.gather(input_tv, dim, index_tv)`. So, the first input is `input_tv`, second is `dim`, and third is `index_tv`. And in these three parameters, only `input_tv` and `index_tv` is `Tensor (TensorView)`, the `dim` is only an `Integer`. And the `input()` method can only get tensor parameters. So `input(0)` is `input_tv`, `input(1)` is `index_tv` and `input(2)` does not make sense.

File: torch/csrc/jit/codegen/cuda/lower_index.cpp (hunk 10 above)
Comment (ftxj) [reply torch/csrc/jit/codegen/cuda/lower_index.cpp]: Thanks. Got it.

File: torch/csrc/jit/codegen/cuda/lower_index.cpp (hunk 11 above)
Comment (ftxj) [reply torch/csrc/jit/codegen/cuda/lower_index.cpp]: Yes, repeated here.

File: torch/csrc/jit/codegen/cuda/lower_index.cpp (hunk 12 above)
Comment (ftxj) [reply torch/csrc/jit/codegen/cuda/lower_index.cpp]: The domain contains the vector seam like `[T0, T1, T2]`. We want to get the vector seems like `[T2 * T1 * T0, T2 * T1, T2]`. If we do not reverse iteration order, we will get `[T2 * T1 * T0, T1 * T0, T0]`. And if we do not reverse the domain on `strides` object. We will get `[T2, T1 * T2, T0 * T1 * T2]`

File: torch/csrc/jit/codegen/cuda/lower_index.cpp (hunk 13 above)
Comment (ftxj) [reply torch/csrc/jit/codegen/cuda/lower_index.cpp]: Ok, I'm refactoring some logic here.

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp (hunk 14 above)
Comment (ftxj) [reply torch/csrc/jit/codegen/cuda/root_domain_map.cpp]: This change makes no sense. Some previous modification was not deleted.

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp (hunk 16 above)
Comment (ftxj) [reply torch/csrc/jit/codegen/cuda/scheduler/registry.cpp]: I will rename these.

File: torch/csrc/jit/codegen/cuda/test/test_gpu_gather_ops.cpp (hunk 17 above)
Comment (ftxj) [reply torch/csrc/jit/codegen/cuda/test/test_gpu_gather_ops.cpp:4]: It's not checked, it's just imitating the other test file.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 18]
```diff
@@ -670,7 +670,10 @@ void IndexCompute::run(const LoopIndexing& loop_indexing) {
     // Resolve missing values from permissive map.
     updateIndexMapFromPermissiveMap(expr);
 
-    handle(expr);
+    // some redundant calculation here
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: Please make this comment more specific

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 19]
```diff
@@ -670,7 +670,10 @@ void IndexCompute::run(const LoopIndexing& loop_indexing) {
     // Resolve missing values from permissive map.
     updateIndexMapFromPermissiveMap(expr);
 
-    handle(expr);
+    // some redundant calculation here 
+    // because collectIndexIntoPermissiveMap has already done
+    handle(expr); 
+
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: nit: remove the empty line

File: torch/csrc/jit/codegen/cuda/lower_index.h [hunk 20]
```diff
@@ -77,6 +78,12 @@ class TORCH_CUDA_CU_API IndexLowering : private OptOutConstDispatch {
       Val* dst,
       const std::unordered_map<IterDomain*, Val*>& override_index = {}) const;
 
+  Val* lowerSizeNonEqualSrcIndex(
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_index.h]: Please add a comment on this function

File: torch/csrc/jit/codegen/cuda/index_compute.h [hunk 21]
```diff
@@ -339,6 +339,14 @@ class Index {
       const IndexFromIdGraph& index_from_id_graph);
 
  public:
+  
+  static kir::TensorIndex* getIndexForNonEqualDomains(
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/index_compute.h]: Please add a comment

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 22]
```diff
@@ -1941,6 +1944,55 @@ std::vector<Val*> Index::getPerDimLogicalIndex(
...
+    if(is_overriden) {
+      unknwon_index.push_back(IrBuilder::mulExpr(override_it->second, unknown_strides[dim]));
+    }
+    else {
+      unknwon_index.push_back(IrBuilder::mulExpr(
+        IrBuilder::modExpr(
+          IrBuilder::divExpr(
+            producer_flatten_index, 
+            producer_strides[dim]),  
+          producer_d->extent()
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: I'm unclear what you're trying to do here. I guess `divExpr(producer_flatten_index, producer_strides[dim])` is to grab an index component from the flattened index val? If so, wouldn't it make more sense to do using `producer_indices`? It seems this part is just reversing what you've already done.

File: torch/csrc/jit/codegen/cuda/index_compute.h (hunk 21 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/index_compute.h]: Also, please use more descriptive parameter names

File: torch/csrc/jit/codegen/cuda/test/test_gpu_gather_ops.cpp [hunk 23]
```diff
@@ -0,0 +1,335 @@
...
+#include <torch/torch.h>
+
+namespace torch {
+namespace jit {
+
+using namespace torch::jit::fuser::cuda;
+
+// pass
+TEST_F(NVFuserTest, TorchGatherOpAllDim_CUDA) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/test/test_gpu_gather_ops.cpp]: Our test naming convention is `Fusion***_CUDA`. Please add `Fusion` to the test names

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/test/test_gpu_gather_ops.cpp]: Also, please add a brief comment about what each test is intended to verify

File: torch/csrc/jit/codegen/cuda/test/test_gpu_gather_ops.cpp [hunk 24]
```diff
@@ -0,0 +1,335 @@
...
+      FusionGuard fg(&fusion);
+
+      TensorView* tv1 = makeContigTensor(rank);
+      TensorView* tv_idx = makeContigTensor(rank, DataType::Int);
+      fusion.addInput(tv1);
+      fusion.addInput(tv_idx);
+      auto tv_out = torch_gather(tv1, dim, tv_idx);
+      fusion.addOutput(tv_out);
+
+      std::cout << fusion << std::endl;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/test/test_gpu_gather_ops.cpp]: Please remove debug output

File: torch/csrc/jit/codegen/cuda/test/test_gpu_gather_ops.cpp [hunk 25]
```diff
@@ -0,0 +1,335 @@
...
+#include <torch/csrc/jit/codegen/cuda/test/test_utils.h>
+
+#include <torch/torch.h>
+
+namespace torch {
+namespace jit {
+
+using namespace torch::jit::fuser::cuda;
+
+// pass
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/test/test_gpu_gather_ops.cpp]: What does this comment mean?

File: torch/csrc/jit/codegen/cuda/test/test_gpu_gather_ops.cpp [hunk 26]
```diff
@@ -0,0 +1,335 @@
...
+TEST_F(NVFuserTest, TorchGatherOpAllDim_CUDA) {
+  const int max_dim_size = 64;
+  std::srand(std::time(nullptr));
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/test/test_gpu_gather_ops.cpp]: Please make the tests deterministic. Don't use the current time as a seed.

File: torch/csrc/jit/codegen/cuda/test/test_gpu_gather_ops.cpp [hunk 27]
```diff
@@ -0,0 +1,335 @@
...
+
+      FusionExecutorCache executor_cache(std::move(fusion_ptr));
+      auto cg_outputs = executor_cache.runFusionWithInputs(aten_inputs);
+      testValidate(
+          &fusion, cg_outputs, aten_inputs, {tv_out_ref}, __LINE__, __FILE__);
+    } 
+  }
+}
+
+// will generate two kernels
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/test/test_gpu_gather_ops.cpp]: Unclear what is supposed to be verified in this test

File: torch/csrc/jit/codegen/cuda/index_compute.cpp (hunk 22 above)
Comment (ftxj) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp]: You are right, I do a lot of unnecessary work here. And will fix quickly.

File: torch/csrc/jit/codegen/cuda/ir_utils.cpp [hunk 28]
```diff
@@ -753,22 +765,28 @@ bool isIndexSelectIndicesTv(const TensorView* tv) {
...
-          "Expect dim of index select is smaller than its output root domain size. But got dim = ",
-          dim,
-          " size = ",
-          tv->domain()->getRootDomain().size());
-      return tv->domain()->getRootDomain()[dim];
+bool isTorchGatherLookupTv(const Val* tv) {
+  for (auto expr : tv->uses()) {
+    if (expr->isA<TorchGatherOp>()) {
+      auto idx_sel = expr->as<TorchGatherOp>();
+      if (idx_sel->input(0) == tv) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_utils.cpp]: Please use `lookupTv()`

File: torch/csrc/jit/codegen/cuda/ir_utils.cpp [hunk 29]
```diff
@@ -753,22 +765,28 @@ bool isIndexSelectIndicesTv(const TensorView* tv) {
...
   }
-  return nullptr;
+  return false;
+}
+
+bool isTorchGatherIndicesTv(const Val* tv) {
+  for (auto expr : tv->uses()) {
+    if (expr->isA<TorchGatherOp>()) {
+      auto idx_sel = expr->as<TorchGatherOp>();
+      if (idx_sel->input(1) == tv) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_utils.cpp]: Here, `indexTv()`

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp (hunk 3 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_iostream.cpp]: @ftxj Please change this

**Review by naoyam — APPROVED** [review]
Approving the PR. Just left a few minor comments. Feel free to merge once addressing them. Thanks!

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp [hunk 30]
```diff
@@ -19,7 +19,7 @@ namespace cuda {
 void checkInlineable(const Expr* expr) {
   for (auto input : expr->inputs()) {
     TORCH_CHECK(
-        input->isScalar() ||
+        input->isScalar() || input->isA<kir::TensorIndex>() ||
```
Comment (ftxj) [inline torch/csrc/jit/codegen/cuda/ir_iostream.cpp:22]: In `torch.gather(input, dim, index)` operator, the selected index in `TensorIndex` of `input tensor` is another `TensorIndex`, which generated using `index tensor`. To support print this operator, we need the `TensorIndex` here.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_iostream.cpp:22]: I guess this would allow basically everything in Kernel printed inline. What error did you have without this change?

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp [hunk 31]
```diff
@@ -180,6 +180,41 @@ std::string IndexSelectOp::toInlineString(int indent_size) const {
...
+  addOutput(out);
+  addAttribute(select_id);
+  addAttribute(IrBuilder::create<Attribute<int>>(passkey.ir_container_, dim));
+}
+
+std::string TorchGatherOp::toString(int indent_size) const {
+  std::stringstream ss;
+  indent(ss, indent_size) << output(0)->toString() << "\n";
+  indent_size++;
+  indent(ss, indent_size) << " = gather( ";
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: As I asked before, please use something like `torch_gather` for now.

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 32]
```diff
@@ -553,6 +555,47 @@ TensorView* index_select(TensorView* lookup_tv, int dim, TensorView* index_tv) {
...
+      inp_domain.size(),
+      " non-reduction dims.");
+  std::vector<IterDomain*> out_domain;
+  for (int i = 0; i < idx_domain.size(); ++i) {
+    out_domain.push_back(
+        IterDomainBuilder(idx_domain[i])
+            .iter_type(
+                idx_domain[i]->getIterType() == IterType::Iteration
+                    ? IterType::GatherScatter
+                    : idx_domain[i]->getIterType())
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/arith.cpp:585]: I suppose this is needed for a case where the index tensor has broadcast domains.

Can the selected axis of the index tensor be a broadcast? I think it can. Pinging @zasdfgbnm

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp [hunk 33]
```diff
@@ -1106,6 +1124,28 @@ void ComputeAtRootDomainMapBuilder::handle(GatherOp* op) {
   }
 }
 
+void ComputeAtRootDomainMapBuilder::handle(TorchGatherOp* op) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/root_domain_map.cpp:1127]: Why did you need this? I does make sense, and I think we should need something similar for `select` and `index_select`, but I though we're still fine without this.

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 32 above)
Comment (ftxj) [reply torch/csrc/jit/codegen/cuda/arith.cpp:585]: When one of the dimension sizes of index tensor is equal to one, it will become broadcast domain.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/arith.cpp:585]: Yeah, but then it's a little inconsistent that sometimes we have `GatherScatter` IterDomains but sometimes `Broadcast`. I still believe we really don't need `GatherScatter`, but removing it would be a different cleanup PR.

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp (hunk 33 above)
Comment (ftxj) [reply torch/csrc/jit/codegen/cuda/root_domain_map.cpp:1127]: When `index` tensor has a broadcast domain, and the `torch_gather` will look like `T[iS2, bS3] = torch_gather(input = T[iS0, iS1], dim = 0, index = T[iS2, bS3])`. If we use `mapPointwiseOrReductionOp` to handle `torch_gather`, the `iS1` and `bS3` will miss match.

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 32 above)
Comment (ftxj) [reply torch/csrc/jit/codegen/cuda/arith.cpp:585]: Yeah, I will explore methods that don't require `GatherScatter`.

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp (hunk 30 above)
Comment (ftxj) [reply torch/csrc/jit/codegen/cuda/ir_iostream.cpp:22]: If we don't allow `TensorIndex` to print inline, some IR like `T1[ T2[idx] + ...]` cannot print.

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp (hunk 33 above)
Comment (ftxj) [reply torch/csrc/jit/codegen/cuda/root_domain_map.cpp:1127]: And the miss match will lead to an `Assert` in [this line](https://github.com/csarofeen/pytorch/blob/74b773ec2acd503082165550db0d14beecbee0b1/torch/csrc/jit/codegen/cuda/root_domain_map.cpp#L860)

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp (hunk 30 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_iostream.cpp:22]: But, wouldn't it still print `T1`? Wouldn't it be just not inlined?

Comment (ftxj) [reply torch/csrc/jit/codegen/cuda/ir_iostream.cpp:22]: It just gives me `Printing inline computations involving values other than scalars is not currently supported.` failure without this modification after the `IndexLowering` pass.

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp (hunk 33 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/root_domain_map.cpp:1127]: OK, that makes sense.

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp (hunk 30 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_iostream.cpp:22]: Oh, I thought `checkInlineable` returns a bool indicating if it's possible to inline. Didn't realize it just fails if not.

What's the command to reproduce the error?

Comment (ftxj) [reply torch/csrc/jit/codegen/cuda/ir_iostream.cpp:22]: Just remove this modification and run `PYTORCH_NVFUSER_DUMP=cuda_kernel,fusion_ir,lower_verbose ./build/bin/test_jit --gtest_filter='NVFuserTest.FusionTorchGatherIndexTvExtentIsOne_CUDA' `

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_iostream.cpp:22]: I think what @ftxj is doing here makes sense, in the `toString` method of `TensorIndex`, we have
```C++
std::string TensorIndex::toString(int indent_size) const {
  std::stringstream ss;
  ss << ir_utils::varName(this);
  switch (view()->getMemoryType()) {
    case MemoryType::Global:
      ss << "_g";
      break;
    case MemoryType::Shared:
      ss << "_s";
      break;
    case MemoryType::Local:
      ss << "_l";
      break;
    default:
      TORCH_INTERNAL_ASSERT(false, "Unknown tensor memory type.");
  }
  ss << "[";
  ss << index()->toInlineString(indent_size);
  ss << "]";
  ss << " view( " << ir_utils::varName(view()) << " )";
  return ss.str();
}
```
which prints the index inline. Without this change, we won't be able to print a `TensorIndex` object.
